### PR TITLE
20.7 fb elisa vis 7808

### PR DESCRIPTION
**Pull request by labkey-klum**
#### Rationale
This is a follow on story to the initial [ELISA assay improvements](https://docs.google.com/document/d/131Mh8D1HMs89YltSr263w931cbsqBi96B--xmTVu4OU/edit) which added support for multi-plate, multi-well data formats. In this story the run details page is reworked to support additional filtering capabilities on:
- plate
- analyte
- samples
- controls

The previous ExtJS based view has been replaced with a new React based view.

#### Related Pull Requests
https://github.com/LabKey/platform/pull/1628

#### Changes
- allow passing XY pairs of plot data to the visualization API
- React based run details view for the ELISA assay
- store curve fit parameters as a JSON text string in a new protocol schema table : CurveFit

### CODE REVIEW

File: elisa/resources/schemas/dbscripts/postgresql/elisa-0.000-20.000.sql [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    RunId INTEGER NOT NULL,
+    ProtocolId INTEGER NOT NULL,
+    PlateName VARCHAR(250),
+    Spot INTEGER,
+    RSquared DOUBLE PRECISION,
+    FitParameters VARCHAR(500),
+
+    CONSTRAINT PK_CurveFit PRIMARY KEY (RowId),
+    CONSTRAINT FK_CurveFit_ProtocolId FOREIGN KEY (ProtocolId) REFERENCES exp.Protocol (RowId),
+    CONSTRAINT FK_CurveFit_RunId FOREIGN KEY (RunId)
```
Comment (cnathe) [inline elisa/resources/schemas/dbscripts/postgresql/elisa-0.000-20.000.sql:36]: Any reason to CREATE INDEX for either of these FKs?

File: elisa/resources/schemas/elisa.xml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+
+ -->
+<tables xsi:schemaLocation="http://labkey.org/data/xml ../../../../../schemas/tableInfo.xsd"
+        xmlns="http://labkey.org/data/xml" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+    <table tableName="CurveFit" tableDbType="TABLE">
+        <description>Contains one row per notebook or experiment in this folder.</description>
```
Comment (cnathe) [inline elisa/resources/schemas/elisa.xml]: looks like this might be a copy-paste error from a different xml file.

**Review by cnathe — APPROVED** [review]
Approved server side code changes. Pending review of client side code from Marty.

File: elisa/src/org/labkey/elisa/ElisaProviderSchema.java [hunk 3]
```diff
@@ -17,25 +17,6 @@
 {
     public static final String CURVE_FIT_METHOD_TABLE_NAME = "CurveFitMethod";
 
-    enum ElisaCurveFits
-    {
-        FOUR_PARAMETER(StatsService.CurveFitType.FOUR_PARAMETER.getLabel()),
-        FIVE_PARAMETER(StatsService.CurveFitType.FIVE_PARAMETER.getLabel()),
```
Comment (cnathe) [inline elisa/src/org/labkey/elisa/ElisaProviderSchema.java:23]: @labkey-klum  Any reason we need/want to leave the 5PL type around for the single plate assay case? I believe it was their prior to this work. I have a local 20.7 MSSQL DB which has a run that had Curve Fit Method = Five Parameter. The upgraded/populated CurveFit table for that container looks like this:
Screen Shot 2020-10-15 at 11 47 41 AM

File: elisa/resources/schemas/dbscripts/postgresql/elisa-0.000-20.000.sql (hunk 1 above)
Comment (labkey-klum) [reply elisa/resources/schemas/dbscripts/postgresql/elisa-0.000-20.000.sql:36]: yup, good point

File: elisa/src/org/labkey/elisa/ElisaProviderSchema.java (hunk 3 above)
Comment (labkey-klum) [reply elisa/src/org/labkey/elisa/ElisaProviderSchema.java:23]: I decided to remove it because it didn't yield good fits. I think if we wanted to add a 5PL fit in the future we would use the new 4PL simplex algorithm.

Comment (cnathe) [reply elisa/src/org/labkey/elisa/ElisaProviderSchema.java:23]: That sounds good. Just checking. Thanks.

File: elisa/test/styleMock.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+
+
+export {};
```
Comment (labkey-martyp) [inline elisa/test/styleMock.ts:3]: Is this file necessary?

File: elisa/src/client/RunDetails/components/PlotButtonBar.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    runId: number,
+    runPropertiesRow: {[key: string]: any},
+    plotOptions: PlotOptions,
+    plotId: string
+}
+
+export class PlotButtonBar extends PureComponent<Props> {
+
+    getSVGElement() {
+        return $('#' + this.props.plotId + '>svg');
```
Comment (labkey-martyp) [inline elisa/src/client/RunDetails/components/PlotButtonBar.tsx]: Like mentioned above, wonder if this could be done with React.createRef instead of using jQuery

File: elisa/src/client/RunDetails/components/CurveFitPanel.tsx [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                            <input
+                                type="checkbox"
+                                name='showCurve'
+                                checked={plotOptions.showCurve}
+                                onChange={this.onShowCurveCheck}
+                            />
+                            Show curve fit line
+                            <LabelHelpTip
+                                title={'Standard Curve Fit'}
+                                body={() => (
```
Comment (labkey-martyp) [inline elisa/src/client/RunDetails/components/CurveFitPanel.tsx]: Looks like LabelHelpTip is a PureComponent.  By redeclaring the function here with every render it's causing the body prop to always be different and unnecessary re-renders of LabelHelpTip

File: elisa/src/client/RunDetails/components/CalibrationCurvePanel.tsx [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        return $('.plot-panel-section').width();
+    }
+
+    renderPlot() {
+        const { runPropertiesRow, plotOptions, data, curveFitData, columnInfo } = this.props;
+        const LABKEY = getServerContext();
+
+        this.getPlotElement().html('<div class="loading-msg"><i class="fa fa-spinner fa-pulse"></i> Rendering plot...</div>');
+
+        setTimeout(() => {
```
Comment (labkey-martyp) [inline elisa/src/client/RunDetails/components/CalibrationCurvePanel.tsx:49]: This setTimeout seems odd. Is it a re-render issue?

File: elisa/src/client/RunDetails/components/CalibrationCurvePanel.tsx [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        this.renderPlot();
+    }
+
+    getPlotElement() {
+        return $('#' + this.state.plotId);
+    }
+
+    getPlotWidth(): number {
+        return $('.plot-panel-section').width();
+    }
```
Comment (labkey-martyp) [inline elisa/src/client/RunDetails/components/CalibrationCurvePanel.tsx]: Wonder if you could use React Refs and not need the dependency on jQuery

File: elisa/src/client/RunDetails/actions.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+                numberOfPoints: 10000
+            },
+            success: Utils.getCallbackWrapper((data) => resolve(data)),
+            failure: Utils.getCallbackWrapper((error) => reject(error))
+        });
+    });
+}
+
+export function exportSVGToFile(svgEl, format, title: string) {
+    if (svgEl && svgEl.length > 0) {
```
Comment (labkey-martyp) [inline elisa/src/client/RunDetails/actions.ts]: svgEl?.length > 0

File: elisa/src/client/RunDetails/RunDetails.tsx [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+                    <Alert bsStyle={'warning'}>
+                        Warning: the assay design is missing the following fields which may affect the various plotting
+                        features and display of this page: {missingFieldKeys.join(', ')}.
+                    </Alert>
+                }
+                <div className={'row'}>
+                    <div className={'col col-xs-12 col-md-3 run-details-left'}>
+                        <DataSelectionsPanel {...this.state} setPlotOption={this.setPlotOption}/>
+                        <PlotOptionsPanel {...this.props} {...this.state} setPlotOption={this.setPlotOption}/>
+                        <CurveFitPanel {...this.props} {...this.state} setPlotOption={this.setPlotOption}/>
```
Comment (labkey-martyp) [inline elisa/src/client/RunDetails/RunDetails.tsx:237]: Along the lines of the comment above, you might be be able to factor out some common interfaces used as props/state for these components and RunDetails.

File: elisa/src/client/RunDetails/RunDetails.tsx [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+        return (
+            <>
+                {missingFieldKeys?.length > 0 &&
+                    <Alert bsStyle={'warning'}>
+                        Warning: the assay design is missing the following fields which may affect the various plotting
+                        features and display of this page: {missingFieldKeys.join(', ')}.
+                    </Alert>
+                }
+                <div className={'row'}>
+                    <div className={'col col-xs-12 col-md-3 run-details-left'}>
```
Comment (labkey-martyp) [inline elisa/src/client/RunDetails/RunDetails.tsx]: I might be misunderstanding the .col class but thought that was more for when you don't have a specific col size class (col-xs-12) on a column.

File: elisa/src/client/RunDetails/RunDetails.tsx [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+    }
+
+    componentDidUpdate(prevProps: Readonly<ImplProps>, prevState: Readonly<ImplState>, snapshot?: any): void
+    {
+        if (shouldReloadCurveFitData(prevState.plotOptions, this.state.plotOptions)) {
+            this.getCurveFitData();
+        }
+    }
+
+    getCurveFitData() {
```
Comment (labkey-martyp) [inline elisa/src/client/RunDetails/RunDetails.tsx]: arrow function and return type?

File: elisa/src/client/RunDetails/RunDetails.tsx [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+}
+
+interface ImplProps {
+    protocolId: number,
+    runId: number,
+    runPropertiesRow: {[key: string]: any},
+    data: any[],
+    columnInfo: {[key: string]: any},
+    measures: string[]
+}
```
Comment (labkey-martyp) [inline elisa/src/client/RunDetails/RunDetails.tsx]: You could carve out a common interface or type between these props and the state for the RunDetails.  Removes some duplicate code and makes sure they're in sync.

File: elisa/src/client/RunDetails/RunDetails.tsx [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+                this.setState(() => ({ dataError: reason.exception }));
+            }
+        });
+    }
+
+    render() {
+        const { runPropertiesError, dataError, runPropertiesRow, data } = this.state;
+
+        if (runPropertiesError || dataError) {
+            return <Alert>{runPropertiesError || dataError}</Alert>
```
Comment (labkey-martyp) [inline elisa/src/client/RunDetails/RunDetails.tsx]: Not sure if possible on the server-side but looking at client side code looks like there could be both of these errors.  Only show one?

File: elisa/src/client/RunDetails/RunDetails.tsx [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+                }
+            },
+            failure: (reason) => {
+                console.error(reason);
+                this.setState(() => ({ runPropertiesError: reason.exception }));
+            }
+        });
+    }
+
+    queryData() {
```
Comment (labkey-martyp) [inline elisa/src/client/RunDetails/RunDetails.tsx]: This is called in a callback in a callback, seems like it should be arrow function to ensure scope.  And missing return type.

**Review by labkey-martyp — COMMENTED** [review]
Looks good Cory.  A handful of comments and questions below.  I did not test this out, comments just based on looking over the code. One general comment, there are a lot of missing return types on functions.

File: elisa/test/styleMock.ts (hunk 4 above)
Comment (cnathe) [reply elisa/test/styleMock.ts:3]: Not at this time, but I decided to keep it in to be consistent with other module jest setups.

File: elisa/src/client/RunDetails/components/CurveFitPanel.tsx (hunk 6 above)
Comment (cnathe) [reply elisa/src/client/RunDetails/components/CurveFitPanel.tsx]: done

File: elisa/src/client/RunDetails/components/PlotButtonBar.tsx (hunk 5 above)
Comment (cnathe) [reply elisa/src/client/RunDetails/components/PlotButtonBar.tsx]: Good call, done. We should likely switch the biologics usage like this at some point as well.

File: elisa/src/client/RunDetails/components/CalibrationCurvePanel.tsx (hunk 7 above)
Comment (cnathe) [reply elisa/src/client/RunDetails/components/CalibrationCurvePanel.tsx:49]: It is just to help the plot refresh slow down a bit and look more natural so that it briefly show the user the "Rendering plot..." message. Without it, the plot just sort of flashes and is updated, so this looks better in my opinion.

File: elisa/src/client/RunDetails/components/CalibrationCurvePanel.tsx (hunk 8 above)
Comment (cnathe) [reply elisa/src/client/RunDetails/components/CalibrationCurvePanel.tsx]: Yes, done.

File: elisa/src/client/RunDetails/actions.ts (hunk 9 above)
Comment (cnathe) [reply elisa/src/client/RunDetails/actions.ts]: With the switch to React.createRef(), this is no longer an array so the if statement changed to just a null check.

File: elisa/src/client/RunDetails/RunDetails.tsx (hunk 11 above)
Comment (cnathe) [reply elisa/src/client/RunDetails/RunDetails.tsx]: You are right, the 'col' isn't needed here. Removed.

File: elisa/src/client/RunDetails/RunDetails.tsx (hunk 12 above)
Comment (cnathe) [reply elisa/src/client/RunDetails/RunDetails.tsx]: It doesn't need to be an arrow function as it is only called from componentDidMount and componentDidUpdate. I'll add the void return type.

File: elisa/src/client/RunDetails/RunDetails.tsx (hunk 13 above)
Comment (cnathe) [reply elisa/src/client/RunDetails/RunDetails.tsx]: I split out a CommonRunProps interface to extend in a few places.

File: elisa/src/client/RunDetails/RunDetails.tsx (hunk 10 above)
Comment (cnathe) [reply elisa/src/client/RunDetails/RunDetails.tsx:237]: I factored out a CommonRunProps interface to extend in a few places.

File: elisa/src/client/RunDetails/RunDetails.tsx (hunk 14 above)
Comment (cnathe) [reply elisa/src/client/RunDetails/RunDetails.tsx]: Done

File: elisa/src/client/RunDetails/RunDetails.tsx (hunk 15 above)
Comment (cnathe) [reply elisa/src/client/RunDetails/RunDetails.tsx]: Done

**Review by labkey-martyp — APPROVED** [review]
